### PR TITLE
feat(plugins): typed invoke arguments for mobile plugins

### PR DESCRIPTION
**Pull request by lucasfernog**
### What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Docs
- [ ] New Binding issue #___
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

### Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

### Checklist
- [ ] When resolving issues, they are referenced in the PR's title (e.g `fix: remove a typo, closes #___, #___`)
- [x] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).
- [ ] I have added a convincing reason for adding this feature, if necessary

### Other information

This PR changes the plugin command API to enforce typed arguments instead of relying on on-demand argument processing via `invoke.getString`, `invoke.getInt`, `invoke.getObject` etc.

New API for Android:

```kotlin
class Options(val recursive: Boolean = false)

class ExampleArgs(val path: String?, val options: Options?, val enabled: Boolean = true)

@TauriPlugin
class ExamplePlugin(private val activity: Activity): Plugin(activity) {
    @Command
    fun run(invoke: Invoke) {
        val args = invoke.parseArgs(ExampleArgs::class.java)
        invoke.resolveObject(args)
    }
}

```

New API for iOS:

```swift
class Options: Decodable {
  let recursive: Bool = false
}

class ExampleArgs: Decodable {
  let path: String?
  let enabled: Bool = true
  let options: Options?
}

class ExamplePlugin: Plugin {
  @objc public func run(_ invoke: Invoke) throws {
    let args = try invoke.parseArgs(ExampleArgs.self)
    invoke.resolve(args)
  }
}
```

Note that you can also resolve with an object instead of the generic `JSObject` class / dictionary.

### CODE REVIEW

**Comment by lucasfernog** [comment]
seems like the jackson dependency costs 1MB on the AAB size, maybe we can optimize that or find a better package for this (though AFAIK there's no Java expert on the team :joy: )

File: core/tauri/mobile/android/src/main/java/app/tauri/plugin/Plugin.kt [hunk 1]
```diff
@@ -167,7 +185,8 @@ abstract class Plugin(private val activity: Activity) {
 
       // If call was made with a list of specific permission aliases to request, save them
       // to be requested
-      val providedPerms: JSArray = invoke.getArray("permissions", JSArray())
+      // TODO val providedPerms: JSArray = invoke.getArray("permissions", JSArray())
```
Comment (amrbashir) [inline core/tauri/mobile/android/src/main/java/app/tauri/plugin/Plugin.kt]: Why was this changed to TODO? should it be solved in this PR?

**Review by amrbashir — DISMISSED** [review]
Android sounds OK to me

File: core/tauri/mobile/android/src/main/java/app/tauri/plugin/Plugin.kt (hunk 1 above)
Comment (lucasfernog) [reply core/tauri/mobile/android/src/main/java/app/tauri/plugin/Plugin.kt]: oops missed this one, thanks for catching it @amrbashir

**Review by amrbashir — APPROVED** [review]

**Comment by amrbashir** [comment]
> seems like the jackson dependency costs 1MB on the AAB size, maybe we can optimize that or find a better package for this (though AFAIK there's no Java expert on the team 😂 )

Java yikes, do these developers exist any more :joy:?
